### PR TITLE
feat(discv5): add crate for interfacing reth network and sigp/discv5

### PR DESCRIPTION
**Pull request by emhane**
- Makes sigp/discv5 usable in reth network by implementing minimal interface `HandleDiscovery`. 
- Parses enodes into multiaddresses (used by CL) that can be added into sigp/enr via public api. 
- Exposes sigp/discv5 api for filtering queries with custom filters, to narrow down lookup queries. this is necessary since mostly eth2 peers are found on discv5 rn. populating kbuckets with eth2 nodes will lead to almost inability to find EL peers if kbuckets become full, since by Kademlia, kbuckets favour long lived connections - this means it will keep the first nodes it puts in its buckets if those nodes stay online.

### CODE REVIEW

File: crates/net/discv5/src/config.rs
Comment (Rjected) [inline crates/net/discv5/src/config.rs:289]: these should just be regular consts, with an enum. Given optimism has added an entry, I understand that we might want to abstract this in the future, but for now I'd prefer this to be simpler

File: crates/net/discv5/src/config.rs
Comment (Rjected) [inline crates/net/discv5/src/config.rs:258]: noting that this is rlpx, not usually called mempool
```suggestion
    /// Returns the RLPx (TCP) socket contained in the [`discv5::Config`]. This socket will be
```

File: crates/net/discv5/src/metrics.rs
Comment (Rjected) [inline crates/net/discv5/src/metrics.rs:14]: why do we need a trait for this? we should be able to include metrics structs inside of components

File: crates/net/discv5/src/config.rs
Comment (Rjected) [inline crates/net/discv5/src/config.rs:288]: so opstack actually includes this key in enrs?

File: crates/net/discv5/src/config.rs
Comment (Rjected) [inline crates/net/discv5/src/config.rs:217]: instead of making this generic, let's just include by default a hashset containing the chains that will be disallowed. If we need a "must include" set, this can be added as another hashset (with mutual exclusion checks when building the config) if the feature is really needed

**Review by Rjected — CHANGES_REQUESTED** [review]
I'm still a bit confused by the downgrade logic, why do we have to include discv4 in this discv5 logic? It seems to make everything more complex, and I'm not sure what the benefit is

File: crates/net/discv5/src/config.rs
Comment (emhane) [reply crates/net/discv5/src/config.rs:288]: some yeah, for example
```
2024-03-22T11:29:19.139714Z TRACE net::discovery::discv5: discovered peer on opstack fork_id=None enr=Enr { id: Some("v4"), seq: 1708293705211, NodeId: 0x8aa3a341cc388b03310b4b13f641722dd2bf0d6a992c1a9bf984489674ef4876, signature: "98ac9fdef036164e4566b84efea47b3f6944f96d8bbef0026830d27346f4253e4731288dddbe4af11dfd5481cfcb6ebc71a56c83dd931a547e7da0a1b75b6a21", IpV4 UDP Socket: Some(78.153.130.198:9222), IpV6 UDP Socket: None, IpV4 TCP Socket: Some(78.153.130.198:9222), IpV6 TCP Socket: None, Other Pairs: [("opstack", "85f1dbda0300"), ("secp256k1", "a1031b9dda693d646e1dabff64e3140c65afeb9c16c6847d40ce8f8936e6b347d496")] }
2024-03-22T11:29:19.140715Z TRACE net::swarm: adding new peer to network peer_id="0x1b9d…4d41" socket_addr=78.153.130.198:9222 fork_id=None
```
not 100% sure if its fork id, or smthg else?

Comment (Rjected) [reply crates/net/discv5/src/config.rs]: hmmm, yeah probably a forkid / forkhash

**Comment by emhane** [comment]
blocked by https://github.com/paradigmxyz/reth/pull/7328, missing `TryFrom<Enr>` impl

File: crates/net/discv5/src/config.rs
Comment (emhane) [reply crates/net/discv5/src/config.rs]: but maybe only for base?

**Comment by emhane** [comment]
re-request review @Rjected

File: crates/net/discv5/src/metrics.rs
Comment (Rjected) [inline crates/net/discv5/src/metrics.rs:9]: should rename to `Discv5Metrics` or something else, the `Metrics` name is overloaded

File: crates/net/discv5/src/lib.rs
Comment (Rjected) [inline crates/net/discv5/src/lib.rs:68]: also not confident in this trait, these could be public methods on a struct and is implemented only once

File: crates/net/discv5/src/config.rs
Comment (Rjected) [inline crates/net/discv5/src/config.rs:36]: let's make these arrays instead of strings

File: crates/net/discv5/src/lib.rs
Comment (Rjected) [inline crates/net/discv5/src/lib.rs:132]: does this need to be pub with hidden docs? would like to make private and remove `#[doc(hidden)]` if possible

File: crates/net/discv5/README.md
Comment (Rjected) [inline crates/net/discv5/README.md:5]: doc should be updated since afaik there is no downgrade logic in this PR

**Review by Rjected — CHANGES_REQUESTED** [review]
took another pass, looks better! I have some comments on traits and lots of random nits

File: crates/net/discv5/src/config.rs
Comment (emhane) [reply crates/net/discv5/src/config.rs:329]: oh fudge, had done that but must have slipt fixing merge conflicts when porting changes from og branch

File: crates/net/discv5/src/config.rs
Comment (emhane) [reply crates/net/discv5/src/config.rs:223]: yeah ok, makes sense. on go branch this was set to the default discv4 address...as that is what network crate does. will require port, like you say.

File: crates/net/discv5/src/lib.rs
Comment (emhane) [reply crates/net/discv5/src/lib.rs:177]: tamam

File: crates/net/discv5/src/lib.rs
Comment (emhane) [reply crates/net/discv5/src/lib.rs:341]: can move to a function `spawn_populate_kbuckets` ?

File: crates/net/discv5/src/lib.rs
Comment (emhane) [reply crates/net/discv5/src/lib.rs:68]: we-hates-it

File: crates/net/discv5/src/lib.rs
Comment (Rjected) [reply crates/net/discv5/src/lib.rs:341]: makes sense, that would be nice

File: crates/net/discv5/src/lib.rs
Comment (Rjected) [inline crates/net/discv5/src/lib.rs:688]: can we use existing U256?

File: crates/net/discv5/src/lib.rs
Comment (Rjected) [inline crates/net/discv5/src/lib.rs:542]: can you explain why this is needed? is this specced somewhere?

File: crates/net/discv5/src/lib.rs
Comment (Rjected) [inline crates/net/discv5/src/lib.rs:158]: let's move this to the top of the file, I prefer to keep imports out of functions

**Review by Rjected — CHANGES_REQUESTED** [review]
this is looking pretty good, I have some smol nits and a question

File: crates/primitives/src/chain/spec.rs
Comment (mattsse) [inline crates/primitives/src/chain/spec.rs:741]: needs #Panics section

File: crates/net/discv5/src/config.rs
Comment (mattsse) [inline crates/net/discv5/src/config.rs:287]: order

File: crates/net/discv5/src/enr.rs
Comment (mattsse) [inline crates/net/discv5/src/enr.rs:10]: this will panic for non-secp256k1?

File: crates/net/discv5/Cargo.toml
Comment (mattsse) [inline crates/net/discv5/Cargo.toml:51]: do we need this?

File: crates/net/discv5/src/lib.rs
Comment (mattsse) [inline crates/net/discv5/src/lib.rs:2]: this needs the regular crate header

File: crates/net/discv5/src/lib.rs
Comment (mattsse) [inline crates/net/discv5/src/lib.rs:273]: I think all this does is just sending a message over a channel and await the response?
I don't think we need to spawn this as a task?

File: crates/net/discv5/src/lib.rs
Comment (mattsse) [inline crates/net/discv5/src/lib.rs:46]: do we need this to be mut?

File: crates/net/discv5/src/lib.rs
Comment (mattsse) [inline crates/net/discv5/src/lib.rs:99]:
```suggestion
        let mut buf = Vec::new();
```

**Review by mattsse — CHANGES_REQUESTED** [review]
this looks good

some nits

File: crates/net/discv5/src/lib.rs
Comment (emhane) [reply crates/net/discv5/src/lib.rs:688]: I just wanted to copy paste the code exactly as it is in sigp, makes easier to follow along than if its some adaptation

File: crates/net/discv5/src/lib.rs
Comment (emhane) [reply crates/net/discv5/src/lib.rs:542]: not specced, it's an addition based on experience. it's better than looking up a random node id in queries that have aim to populate kbuckets. for a random node id, there is 50% chance the node id falls in the bucket furthest away. as requested distances is capped at 3 distances, the likelihood that any peers discovered as part of a query, will fall in the top 3 kbuckets is 50%. that means all peers discovered in all queries, are competing over the same 16 * 3 places in the kbuckets, where 16 is the max nodes per kbuckets. for health of the network we want to be connected to every subtree in the network.
Screenshot 2024-04-01 at 22 04 56

Comment (emhane) [reply crates/net/discv5/src/lib.rs:542]: note: if all apps using discv5 did this, DAS would be a piece of cake, which is about finding specific node ids, because nodes that are online for a longer time, would have nicely populated kbuckets (not only the 3 buckets furthest away). I think we want to keep this bigger perspective for what's beneficial for whole ethereum now when we enter discv5 universe :)

File: crates/net/discv5/src/enr.rs
Comment (emhane) [reply crates/net/discv5/src/enr.rs:10]: yes but that is caught already by discv5 on decoding level, the key won't decode if it's another type. so here in the reth code, it's infallible.

https://github.com/sigp/discv5/blob/6707bcf3da350ff273208bb38dc29acc4cc4850c/src/rpc.rs#L459-L460
https://github.com/sigp/enr/blob/83cd98990e7f191b310addb202869994a8f59c96/src/keys/k256_key.rs#L56-L60

File: crates/net/discv5/src/lib.rs
Comment (emhane) [reply crates/net/discv5/src/lib.rs:273]: it makes the requests run in parallel

File: crates/net/discv5/Cargo.toml
Comment (mattsse) [reply crates/net/discv5/Cargo.toml:47]: the uint, why are we defining new uint types?

File: crates/net/discv5/src/enr.rs
Comment (mattsse) [reply crates/net/discv5/src/enr.rs:10]: then this function should be pub crate

File: crates/net/discv5/src/filter.rs
Comment (mattsse) [inline crates/net/discv5/src/filter.rs:50]: why is this a dashset?

File: crates/net/discv5/src/lib.rs
Comment (mattsse) [reply crates/net/discv5/src/lib.rs:277]: they get all sent to the same service.
doing this on separate tasks is unnecessary.

File: crates/net/discv5/src/lib.rs
Comment (mattsse) [reply crates/net/discv5/src/lib.rs:691]: we already have an u256 type

**Review by mattsse — CHANGES_REQUESTED** [review]
few more nits

File: crates/net/discv5/src/filter.rs
Comment (emhane) [reply crates/net/discv5/src/filter.rs:50]: I was sending this filter to sigp/discv5, but don't need it anymore

**Comment by emhane** [comment]
> the uint, why are we defining new uint types?

because I copied a module from sigp/discv5, because I didn't want to invent smthg new to test the code, i wanted to use code that is used in production and is heavily audited. it's just a dev dep so I'm hoping it doesn't disturb you.

> then this function should be pub crate

related to the point above, if libraries would just make functions public, code would seldom have to be copy pasted. I'm relecetuant to restrict visibility since I don't know in what way users may want to use reth, and want it to be as fun as possible for them. an inaccessible function doesn't spark joy.

> they get all sent to the same service.
doing this on separate tasks is unnecessary.

yes, but they only get sent when the preceding query finishes if not spawned on separate threads.

File: crates/net/discv5/src/enr.rs
Comment (emhane) [reply crates/net/discv5/src/enr.rs:10]: I'll add a panic section

**Comment by Rjected** [comment]
> > the uint, why are we defining new uint types?
> 
> because I copied a module from sigp/discv5, because I didn't want to invent smthg new to test the code, i wanted to use code that is used in production and is heavily audited. it's just a dev dep so I'm hoping it doesn't disturb you.

let's remove the `construct_uint` call and use our `U256`, because `Key` logic is more important wrt testing imo

**Comment by emhane** [comment]
> > > the uint, why are we defining new uint types?
> > 
> > 
> > because I copied a module from sigp/discv5, because I didn't want to invent smthg new to test the code, i wanted to use code that is used in production and is heavily audited. it's just a dev dep so I'm hoping it doesn't disturb you.
> 
> let's remove the `construct_uint` call and use our `U256`, because `Key` logic is more important wrt testing imo

alright sure, how do I do this with our type (need to re-write the copied sigp module so I can run the test)?
```rust
let a = U256::from(self.hash.as_slice());
```

**Review by mattsse — APPROVED** [review]
I think this is a great standalone service and this should give us everything we need for reth-network

lgtm, pending @Rjected

**Review by Rjected — APPROVED** [review]
LGTM! this should unblock a lot of things

**Pull request merged**
